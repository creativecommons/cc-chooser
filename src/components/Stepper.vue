<template>
    <div class="stepper-container column">
        <div
            v-for="(step, idx) in visibleSteps()"
            :key="idx"
            :class="['step-container', step.name, step.status, enabledQualifier(step.enabled)]"
        >
            <div
                :class="['step-header']"
                @click="setActiveStep(step.id)"
            >
                <h5 class="step-title vocab hb h5b">
                    {{ $t(stepHeaderText(step.name, step.status)) }}
                </h5>
            </div>
            <FirstStep
                v-if="step.status!=='inactive' && step.name==='FS'"
                :step-id="step.id"
                :selected="step.selected"
                :status="step.status"
                @change="changeStepSelected"
            />
            <Step
                v-else-if="step.status!=='inactive' && isLicenseAttribute(step.name)"
                :step-id="step.id"
                :step-name="step.name"
                :selected="step.selected"
                :status="step.status"
                :reversed="isStepReversed(step.name)"
                :enabled="step.enabled"
                :disabled-due="step.disabledDue"
                @change="changeStepSelected"
            />
            <CopyrightWaiverStep
                v-else-if="step.status!=='inactive' && step.name==='CW'"
                :step-id="step.id"
                :step-name="step.name"
                :selected="step.selected"
                :status="step.status"
                @change="changeStepSelected"
            />
            <DropdownStep
                v-else-if="step.status!=='inactive' && step.name==='DD'"
                :step-id="step.id"
                :status="step.status"
                @change="changeStepSelected"
            />
            <AttributionDetailsStep
                v-else-if="step.status!=='inactive' && step.name==='AD'"
                :step-id="step.id"
                :status="step.status"
            />
            <nav
                v-if="step.status==='current'"
                class="step-navigation"
            >
                <a
                    v-if="step.name!=='FS'"
                    role="button"
                    class="pagination-previous"
                    @click="handlePrevious(step.name)"
                >{{ $t('stepper.nav.previous-label') }}</a>
                <a
                    v-if="step.name!=='AD'"
                    role="button"
                    :class="['pagination-next', nextButtonEnabledState(step.id)]"
                    @click="handleNext(step.name)"
                >{{ $t('stepper.nav.next-label') }}</a>
                <span
                    v-else
                    class="pagination-finish"
                >{{ $t('stepper.nav.finish-label') }}</span>
            </nav>
        </div>
    </div>
</template>

<script>
import Step from './Step'
import FirstStep from './FirstStep'
import AttributionDetailsStep from './AttributionDetailsStep'
import CopyrightWaiverStep from './CopyrightWaiverStep'
import DropdownStep from './DropdownStep'
import { updateVisibleEnabledStatus } from '../utils/license-utilities'

export default {
    name: 'Stepper',
    components: {
        FirstStep,
        Step,
        AttributionDetailsStep,
        CopyrightWaiverStep,
        DropdownStep
    },
    props: {
        value: {
            type: Number,
            default: 0
        }
    },
    data() {
        return {
            /** Data for 7 Stepper steps
             * FS: First step, checks if user knows the license (and we need to open the dropdown)
             * BY, NC, ND, SA: license attribute selection steps
             * DD: Step with Dropdown for quick license selection, opens if user knows the license
             * CW: Copyright waiver step for CC0 if the user selects NO on BY step
             * AD: Attribution Details step with the form
             *
             * Properties:
             * visible: sets whether the step should be shown or not. Eg. if BY is selected,
             * Copyright Waiver should not be shown, as the user will not waive copyright
             *
             * enabled: sets whether the step can be clicked/selected.
             * Eg. SA shouldn't be selectable if ND was selected
             *
             * status: can be set to 'current', 'previous', or 'inactive', to show the user's
             * progress in Stepper
             *
             * selected: set to undefined before the user interacts with a step; true/false after
             * user selects radio buttons/ checkboxes/ etc.
             */
            steps: [
                { id: 0, name: 'FS', visible: true, enabled: true, status: 'current', selected: undefined },
                { id: 1, name: 'BY', visible: true, enabled: true, status: 'inactive', selected: undefined },
                { id: 2, name: 'NC', visible: true, enabled: true, status: 'inactive', selected: undefined },
                { id: 3, name: 'ND', visible: true, enabled: true, status: 'inactive', selected: undefined },
                { id: 4, name: 'SA', visible: true, enabled: true, status: 'inactive', selected: undefined },
                { id: 5, name: 'DD', visible: false, enabled: true, status: 'inactive', selected: undefined },
                { id: 6, name: 'CW', visible: false, enabled: true, status: 'inactive', selected: undefined },
                { id: 7, name: 'AD', visible: true, enabled: true, status: 'inactive', selected: undefined }
            ]
        }
    },
    computed: {
        currentStepId: {
            get() { return this.$props.value },
            set(newVal) {
                this.$emit('input', newVal)
            }
        }
    },
    created: function() {
        /**
         * Updates the steps array when license is changed in store through dropdown selection
         */
        this.$store.subscribe((mutation, state) => {
            if (mutation.type === 'updateAttributesFromShort') {
                for (const step in this.steps) {
                    const stepId = this.steps[step].id
                    const stepName = this.steps[step].name
                    const isStepSelected = this.steps[step].selected
                    const isAttrSelected = state.currentLicenseAttributes[stepName]
                    if (this.isLicenseAttribute(stepName) && isStepSelected !== isAttrSelected) {
                        this.$set(this.steps, stepId, { ...this.steps[stepId], selected: isAttrSelected })
                        this.updateDisabledAndVisibleSteps(stepName, isAttrSelected)
                    }
                }
            }
        })
    },
    methods: {
        /**
         * stepHeader shows step 'question' for current step, and step 'heading' for others
         * @returns {string} key for i18n message
         */
        stepHeaderText(stepId, stepStatus) {
            const prefix = `stepper.${stepId}`
            if (stepId === 'AD') {
                return prefix + '.heading'
            }
            return stepStatus === 'current' ? `${prefix}.question` : `${prefix}.heading`
        },
        enabledQualifier(isEnabled) {
            return isEnabled ? 'enabled' : 'disabled'
        },
        isLicenseAttribute(stepName) {
            return ['BY', 'NC', 'ND', 'SA'].indexOf(stepName) > -1
        },
        isStepReversed(stepName) {
            /**
             * NC, ND and SA steps are reversed: unlike BY, they are selected when the user
             * answers no, and not selected when the user answers yes
             */
            return ['NC', 'ND', 'SA'].indexOf(stepName) > -1
        },
        nextButtonEnabledState(stepId) {
            /**
             * Disables the 'Next' button before the user interacts with the step
             */
            return this.steps[stepId].selected === undefined
                ? 'disabled'
                : ''
        },
        changeStepSelected(stepName, stepId, isSelected) {
            /**
             * When a user interacts with a step, updates:
             * 1. 'selected' property for the step in steps array
             * 2. if the step is for license attribute, its value in Vuex store
             * 3. 'enabled' and 'visible' properties of other steps affected by current step selection
             */
            if (this.isLicenseAttribute(stepName)) {
                this.$store.commit('setSelected', { stepName, isSelected })
                // When the user first selects a license attribute, the dropdown step's Next button should be enabled
                // as the dropdown will be populated with the selected license from the state
                const DROPDOWNSTEP = 5
                if (this.steps[DROPDOWNSTEP].selected === undefined && stepName === 'BY') {
                    this.$set(this.steps, DROPDOWNSTEP, { ...this.steps[DROPDOWNSTEP], selected: true })
                }
            }
            this.$set(this.steps, stepId, { ...this.steps[stepId], selected: isSelected })
            this.updateDisabledAndVisibleSteps()
        },
        handleNext(stepName) {
            /**
             * Updates state when 'Next' button is clicked if current step has been interacted with
             *
             * Finds the next available enabled and visible step, sets its status to 'current',
             * and steps before that to 'previous'
             */
            const stepSelected = this.steps[this.currentStepId].selected
            if (stepSelected === undefined && this.currentStepId <= 6) return
            const nextStep = this.steps.slice(this.currentStepId + 1).find(step => step.visible && step.enabled).id
            this.$set(this.steps, this.currentStepId, { ...this.steps[this.currentStepId], status: 'previous' })
            if (nextStep - this.currentStepId > 1) {
                for (let i = this.currentStepId + 1; i < nextStep; i++) {
                    this.$set(this.steps, i, { ...this.steps[i], status: 'previous', disabledDue: stepName })
                }
            } else {
                this.$set(this.steps, this.currentStepId, { ...this.steps[this.currentStepId], status: 'previous' })
            }
            this.$set(this.steps, nextStep, { ...this.steps[nextStep], status: 'current' })
            this.currentStepId = nextStep
        },
        handlePrevious() {
            /**
             * Updates state when 'Previous' button
             *
             * Finds the previous enabled and visible step, sets its status to 'current',
             * and steps after that to 'inactive'
             */

            let previousStep = this.currentStepId
            for (let i = this.currentStepId - 1; i >= 0; i--) {
                const thisStep = this.steps[i]
                if (thisStep.visible && thisStep.enabled) {
                    previousStep = this.steps[i].id
                    break
                }
            }
            if (this.currentStepId - previousStep > 1) {
                for (let i = this.currentStepId; i > previousStep; i--) {
                    this.$set(this.steps, i, { ...this.steps[i], status: 'inactive' })
                }
            } else {
                this.$set(this.steps, this.currentStepId, { ...this.steps[this.currentStepId], status: 'inactive' })
            }
            this.$set(this.steps, previousStep, { ...this.steps[previousStep], status: 'current' })
            this.currentStepId = previousStep
        },
        setActiveStep(clickedStepId) {
            /**
             * Handles a click on Step header
             */
            if (!this.steps[clickedStepId].enabled) return
            if (this.steps[clickedStepId].status === 'inactive') return
            // only steps before the current one are clickable
            if (clickedStepId >= this.currentStepId) return
            for (let i = this.currentStepId; i > clickedStepId; i--) {
                this.$set(this.steps, i, { ...this.steps[i], status: 'inactive' })
            }
            this.$set(this.steps, clickedStepId, { ...this.steps[clickedStepId], status: 'current' })
            this.currentStepId = clickedStepId
        },
        setStepsVisible(stepsToSetVisible) {
            // sets steps in stepsToSetVisible array visible properties to true
            this.steps.forEach((step) => {
                if (stepsToSetVisible.indexOf(step.name) > -1 && !step.visible) {
                    this.$set(this.steps, step.id, { ...step, visible: true })
                } else if (stepsToSetVisible.indexOf(step.name) === -1 && step.visible) {
                    this.$set(this.steps, step.id, { ...step, visible: false })
                }
            })
        },
        setStepsEnabled(stepsToSetEnabled, disabledDue) {
            // sets steps in stepsToSetDisabled array enabled properties to false
            this.steps.forEach((step) => {
                // step is currently enabled, but should be disabled
                if (stepsToSetEnabled.indexOf(step.name) === -1 && step.enabled) {
                    this.$set(this.steps, step.id, { ...step, enabled: false, disabledDue: disabledDue })
                } else if (stepsToSetEnabled.indexOf(step.name) > -1 && !step.enabled) {
                    // step is currently disabled, but should be set enabled
                    this.$set(this.steps, step.id, { ...step, enabled: true, disabledDue: '' })
                }
            })
        },
        updateDisabledAndVisibleSteps() {
            /**
             * Creates an array of steps that should be visible/enalbed based on data from steps array
             * and updates the steps array
             */
            const stepsStatusData = {}
            this.steps.forEach((step) => {
                stepsStatusData[step.name] = step.selected
            })
            const { visible, enabled, disabledDue } = updateVisibleEnabledStatus(stepsStatusData)
            this.setStepsVisible(visible)
            this.setStepsEnabled(enabled, disabledDue)
        },
        visibleSteps() {
            return this.steps.filter(step => {
                return step.visible
            })
        }
    }
}
</script>

<style lang="scss">
    .step-container {
        background-color: white;
        border: 2px solid  #D8D8D8;
        border-bottom: none;
        border-radius: 4px;
        max-width: 100%;
        position: relative;
        padding-bottom: 8px;
        &:last-of-type {
            border-bottom: 2px solid #D8D8D8;
            margin-bottom: 15rem;
         }
    }
    .step-header {
        background-color: transparent;
        -webkit-box-align: stretch;
        align-items: stretch;
        display: -webkit-box;
        display: flex;
        position:relative;
        margin: 24px 24px 8px;
        &:hover {
            cursor: pointer;
        }
    }

    .step-header .step-title {
        margin-left: 45px;
    }
    .step-header .step-title::before{
        counter-increment: step-counter;
        position: absolute;
        left: 0;
        display: inline-block;
        font-weight: 700;
        font-family: inherit;
        content: counter(step-counter);
        font-size: 16px;
        width: 30px;
        height: 30px;
        line-height: 30px;
        background: #04A635;
        border-radius: 50%;
        text-align: center;
        color: #fff;
        top: 0;
    }
    .previous.disabled .step-title::before,
    .inactive .step-title::before {
        background-color: #D8D8D8;
        color: #333333;
    }
    .step-content {
        margin-left: 69px;
        padding-bottom:8px;
        padding-right: 24px;
    }
    .step-description {
        color: #333333;
    }
    .current,
    .previous {
        color: black;
    }
    .previous.disabled {
        color: #B0B0B0;
    }
    .previous.disabled .step-title,
    .inactive .step-title {
        color: #B0B0B0;
    }
    .step-navigation {
        margin: 13px 0;
        padding-left: 65px;
    }
    .step-navigation .pagination-next,
    .step-navigation .pagination-previous {
        font-family: Roboto Condensed,sans-serif;
        font-style: normal;
        font-weight: 500;
        font-size: 18px;
        line-height: 24px;
    }
    .step-navigation .pagination-previous {
        border: 2px solid #787878;
        color: #787878;
    }
    .step-navigation .pagination-next {
        background-color: #04a635;
        color: #fff!important;
        border: 2px solid transparent;
    }
    .pagination-next.disabled {
        background-color: #D8D8D8;
    }
    .pagination-next.disabled:hover,
    .pagination-next.disabled:active {
        box-shadow: none;
        border: 1px solid transparent;
    }
    .pagination-finish{
        margin-left: 1rem;
        line-height: 42px;
    }
    .slide-enter-active {
        /*transition: all .3s ease;*/
    animation: slide-down .3s;
}
    .slide-leave-active {
        /*transition: all .8s cubic-bezier(1.0, 0.5, 0.8, 1.0);*/
        animation: slide-down .5s reverse;
    }
    /*.slide-enter, .slide-leave-to {*/
    /*    transform: translateY(-50px);*/
    /*    opacity: 0;*/
    /*}*/
    @keyframes slide-down {
        0% {
            opacity: 0;
            transform: translateY(-100px);
        }

        /*50% {*/
        /*    transform: scaleY(0.5);*/
        /*}*/

        100% {
            opacity: 1;
            transform: translateY(0px);
            /*transform: scaleY(1);*/
        }
    }
    @media (max-width: 860px) {
        .step-container:last-of-type {
            margin-bottom: 1rem;
        }
    }
</style>
