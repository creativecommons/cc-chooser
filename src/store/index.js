import Vue from 'vue'
import Vuex from 'vuex'
import { defaultAttributes, CC0Attributes, attrToShort, attrToFull, licenseUrl, licenseIconsArr } from '../utils/license-utilities.js'

Vue.use(Vuex)

const defaultState = {
    currentLicenseAttributes: { ...defaultAttributes },
    attributionDetails: {
        creatorName: '',
        creatorProfileUrl: '',
        workTitle: '',
        workUrl: ''
    }
}

const createStore = (state) => {
    const initialState = { ...defaultState, ...state }
    return new Vuex.Store({
        state: initialState,
        getters: {
            isLicenseSelected: state => {
                /**
                 * By default, all four license attributes are undefined
                 * As soon as the first attribute(BY) is selected (true/false),
                 * we can show the recommended license
                 */
                return state.currentLicenseAttributes.BY !== undefined
            },
            shortName: state => {
                return attrToShort(state.currentLicenseAttributes)
            },
            fullName: state => {
                return attrToFull(state.currentLicenseAttributes)
            },
            licenseUrl: state => (mode) => {
                return licenseUrl(state.currentLicenseAttributes, mode)
            },
            iconsList: state => {
                return licenseIconsArr(state.currentLicenseAttributes)
            }
        },
        mutations: {
            /**
             * Updates current license attributes when user selects radio option.
             * Edge case: If user selects ND, SA should be set to false
             * @param state
             * @param {string} stepName
             * @param {Boolean} isSelected
             */
            setSelected(state, { stepName, isSelected }) {
                if (stepName === 'ND' && isSelected && state.currentLicenseAttributes.SA) {
                    state.currentLicenseAttributes = {
                        ...state.currentLicenseAttributes,
                        SA: false,
                        ND: true
                    }
                } else if (['BY', 'NC', 'ND', 'SA'].indexOf(stepName) > -1) {
                    state.currentLicenseAttributes = {
                        ...state.currentLicenseAttributes,
                        [stepName]: isSelected
                    }
                }
            },
            updateAttributesFromShort(state, shortName) {
                if (shortName.includes('CC0')) {
                    state.currentLicenseAttributes = { ...CC0Attributes }
                } else {
                    state.currentLicenseAttributes.BY = true
                    state.currentLicenseAttributes.NC = !!shortName.includes('NC')
                    state.currentLicenseAttributes.ND = !!shortName.includes('ND')
                    state.currentLicenseAttributes.SA = !!shortName.includes('SA')
                }
            },
            setCreatorName(state, newName) {
                state.attributionDetails.creatorName = newName
            },
            setCreatorProfileUrl(state, newName) {
                state.attributionDetails.creatorProfileUrl = newName
            },
            setWorkTitle(state, newName) {
                state.attributionDetails.workTitle = newName
            },
            setWorkUrl(state, newName) {
                state.attributionDetails.workUrl = newName
            }
        }
    })
}

export default createStore
