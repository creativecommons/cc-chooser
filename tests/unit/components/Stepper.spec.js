import Vuex from 'vuex'
import { shallowMount, createLocalVue, config } from '@vue/test-utils'
import Stepper from '@/components/Stepper'
import Buefy from 'buefy'
import VueI18n from 'vue-i18n'
import Vue from 'vue'
import store from '@/store'

function getNextButton(wrapper) {
    return wrapper.find('.pagination-next')
}
function clickNext(wrapper) {
    getNextButton(wrapper).trigger('click')
}
function getStepId(wrapper, stepName) {
    return wrapper.vm.steps.filter((step) => { return step.name === stepName })[0].id
}
function setStepSelected(wrapper, stepName, isSelected) {
    const stepId = getStepId(wrapper, stepName)
    wrapper.vm.changeStepSelected(stepName, stepId, isSelected)
}
function stepHeadingText(steps, index) {
    return steps.at(index).find('h5').text()
}
function advanceStep(wrapper, actions) {
    for (const { stepName, isSelected, license } of actions) {
        const stepId = getStepId(wrapper, stepName)
        if (stepName === 'DD') {
            wrapper.vm.$store.commit('updateAttributesFromShort', license)
        }
        wrapper.vm.changeStepSelected(stepName, stepId, isSelected)
        wrapper.vm.handleNext(stepName)
    }
}

let wrapper
let localVue

function setUp() {
    localVue = createLocalVue()
    localVue.use(Vuex)
    localVue.use(Buefy)
    Vue.use(VueI18n)
    const messages = require('@/locales/en.json')
    const i18n = new VueI18n({
        locale: 'en',
        fallbackLocale: 'en',
        messages: messages
    })

    config.mocks.i18n = i18n

    config.mocks.$t = key => {
        // key is a string (eg. 'stepper.ND.question')
        // this line converts it into an object reference
        // eg. messages['stepper.ND.question'] -> messages.stepper.ND.question
        return key.split('.').reduce((messages, k) => messages[k], i18n.messages)
    }
    wrapper = shallowMount(Stepper, {
        store,
        localVue
    })
    wrapper.vm.$on('input', (newVal) => {
        wrapper.setProps({ value: newVal })
    })
}

beforeEach(() => setUp())

describe('Stepper.vue', () => {
    describe('renders correctly', () => {
        it('is called', () => {
            expect(wrapper.name()).toBe('Stepper')
            expect(wrapper.isVueInstance()).toBeTruthy()
        })
        it('has expected UI initially', () => {
            expect(wrapper).toMatchSnapshot()
        })
        it('has expected UI on ND step', () => {
            advanceStep(wrapper, [
                { stepName: 'FS', isSelected: false },
                { stepName: 'BY', isSelected: true },
                { stepName: 'NC', isSelected: false }
            ])
            expect(wrapper).toMatchSnapshot()
        })
        it('has expected UI on CW step after DD', () => {
            advanceStep(wrapper, [
                { stepName: 'FS', isSelected: true },
                { stepName: 'DD', isSelected: true, license: 'CC0 1.0' }
            ])
            expect(wrapper).toMatchSnapshot()
        })
    })

    describe('Step headings', () => {
        it('inactive step headings are not clickable', () => {
            const stepHeaders = wrapper.findAll('.step-header')
            const steps = wrapper.findAll('.step-container')
            stepHeaders.at(0).trigger('click')
            expect(steps.at(0).classes('current')).toBe(true)
            stepHeaders.at(1).trigger('click')
            expect(steps.at(1).classes('inactive')).toBe(true)
        })
        it('clicking on disabled previous step does not change current step', () => {
            advanceStep(wrapper,
                [
                    { stepName: 'FS', isSelected: false },
                    { stepName: 'BY', isSelected: false }
                ])
            expect(wrapper.find('.current').classes()).toContain('CW')
            wrapper.findAll('.step-header').at(2).trigger('click')
            expect(wrapper.find('.current').classes()).toContain('CW')
        })
    })
    describe('Next button', () => {
        it('becomes clickable only after selection is made', () => {
            const nextButton = wrapper.find('.pagination-next')
            expect(nextButton.classes('disabled')).toBe(true)
            setStepSelected(wrapper, 'FS', false)
            expect(nextButton.classes('disabled')).toBe(false)
        })
        it('clicking on Next button advances the step', () => {
            setStepSelected(wrapper, 'FS', true)
            clickNext(wrapper)
            const steps = wrapper.findAll('.step-container')
            expect(steps.at(0).classes('previous')).toBe(true)
            expect(steps.at(1).classes('current')).toBe(true)
        })
    })
    describe('Previous button', () => {
        it('AttributionDetails Step opens correct step when Previous button is clicked', () => {
            // For BY- licenses, it should open SA step
            advanceStep(wrapper, [
                { stepName: 'FS', isSelected: false },
                { stepName: 'BY', isSelected: true },
                { stepName: 'NC', isSelected: false },
                { stepName: 'ND', isSelected: false },
                { stepName: 'SA', isSelected: false }])
            wrapper.find('.pagination-previous').trigger('click')
            expect(wrapper.find('.current').classes()).toContain('SA')
            // for CC0 license, it should open CW step
            wrapper.findAll('.step-header').at(1).trigger('click')
            setStepSelected(wrapper, 'BY', false)
            clickNext(wrapper)
            setStepSelected(wrapper, 'CW', true)
            clickNext(wrapper)
            wrapper.find('.pagination-previous').trigger('click')
            expect(wrapper.find('.current').classes()).toContain('CW')
            // if using dropdown with a BY- license, it should open DD step
            wrapper.findAll('.step-header').at(0).trigger('click')
            advanceStep(wrapper, [{ stepName: 'FS', isSelected: true }])
            advanceStep(wrapper, [{ stepName: 'DD', isSelected: true, license: 'CC BY-NC-ND 4.0' }])
            wrapper.find('.pagination-previous').trigger('click')
            expect(wrapper.find('.current').classes()).toContain('DD')
            console.log(wrapper.emitted())
        })
    })
    describe('FirstStep interactions', () => {
        it('choosing Yes sets 3 steps visible: FS, Dropdown and AttributionDetails, opens DD', () => {
            advanceStep(wrapper, [
                { stepName: 'FS', isSelected: true }])
            const steps = wrapper.findAll('.step-container')
            expect(steps.length).toEqual(3)
            expect(steps.at(0).classes('previous')).toBe(true)
            expect(wrapper.find('.current').classes('DD')).toBe(true)
        })
        it('choosing No sets 6 steps visible: FS, BY, NC, ND, SA and AttributionDetails, opens BY', () => {
            setStepSelected(wrapper, 'FS', false)
            clickNext(wrapper)
            const steps = wrapper.findAll('.step-container')
            expect(steps.length).toEqual(6)
            expect(wrapper.vm.currentStepId).toEqual(1)
            expect(stepHeadingText(steps, 1)).toEqual(wrapper.vm.$t('stepper.BY.question'))
        })
    })
    describe('Step interactions', () => {
        it('BY: selecting No updates enabled steps; disables NC, ND and SA; opens CopyrightWaiverStep', () => {
            advanceStep(wrapper, [{ stepName: 'FS', isSelected: false }])
            setStepSelected(wrapper, 'BY', false)
            const nextButton = getNextButton(wrapper)
            expect(nextButton.classes('disabled')).toBe(false)
            const steps = wrapper.findAll('.step-container')
            const disabledSteps = wrapper.findAll('.disabled')
            expect(disabledSteps.length).toEqual(3)
            expect(steps.length).toEqual(7)
            clickNext(wrapper)
            expect(wrapper.find('.current').classes()).toContain('CW')
        })
        it('ND: selecting ND license updates enabled steps; disables SA step; opens AttributionDetailsStep', () => {
            advanceStep(wrapper, [
                { stepName: 'FS', isSelected: false },
                { stepName: 'BY', isSelected: true },
                { stepName: 'NC', isSelected: false }])
            setStepSelected(wrapper, 'ND', true)
            const steps = wrapper.findAll('.step-container')
            const disabledSteps = wrapper.findAll('.disabled')
            expect(disabledSteps.length).toEqual(1)
            expect(steps.length).toEqual(6)
            clickNext(wrapper)
            expect(wrapper.find('.current').classes()).toContain('AD')
        })
    })
    describe('DropdownStep interactions', () => {
        beforeEach(() => {
            // setUp()
            advanceStep(wrapper, [
                { stepName: 'FS', isSelected: true }])
        })
        it('selecting CC0 makes 4 steps visible and opens Copyright Waiver step', async() => {
            const shortName = 'CC0 1.0'
            advanceStep(wrapper, [{ stepName: 'DD', isSelected: true, license: shortName }])
            const steps = wrapper.findAll('.step-container')
            expect(steps.length).toEqual(4)
            expect(wrapper.find('.current').classes()).toContain('CW')
        })
        it('selecting a BY license and clicking Next makes 3 steps visible and opens AttributionDetails step', async() => {
            advanceStep(wrapper, [{ stepName: 'DD', isSelected: true, license: 'CC BY 4.0' }])
            const steps = wrapper.findAll('.step-container')
            expect(steps.length).toEqual(3)
            expect(wrapper.find('.current').classes()).toContain('AD')
        })
    })
})
