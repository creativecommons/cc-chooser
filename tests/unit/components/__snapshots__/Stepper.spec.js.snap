// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper.vue renders correctly has expected UI initially 1`] = `
<div class="stepper-container column">
  <div class="step-container FS current enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Do you know which license you need?
      </h5>
    </div>
    <firststep-stub stepid="0" status="current"></firststep-stub>
    <nav class="step-navigation">
      <!----> <a role="button" class="pagination-next disabled">NEXT STEP</a></nav>
  </div>
  <div class="step-container BY inactive enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Attribution
      </h5>
    </div>
    <!---->
    <!---->
  </div>
  <div class="step-container NC inactive enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Commercial Use
      </h5>
    </div>
    <!---->
    <!---->
  </div>
  <div class="step-container ND inactive enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Derivative Works
      </h5>
    </div>
    <!---->
    <!---->
  </div>
  <div class="step-container SA inactive enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Sharing Requirements
      </h5>
    </div>
    <!---->
    <!---->
  </div>
  <div class="step-container AD inactive enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Attribution Details
      </h5>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`Stepper.vue renders correctly has expected UI on CW step after DD 1`] = `
<div class="stepper-container column">
  <div class="step-container FS previous enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        License Expertise
      </h5>
    </div>
    <firststep-stub stepid="0" status="previous" selected="true"></firststep-stub>
    <!---->
  </div>
  <div class="step-container DD previous enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Creative Commons License
      </h5>
    </div>
    <dropdownstep-stub status="previous" stepid="5"></dropdownstep-stub>
    <!---->
  </div>
  <div class="step-container CW current enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Waive Your Copyright
      </h5>
    </div>
    <copyrightwaiverstep-stub stepid="6" stepname="CW" status="current"></copyrightwaiverstep-stub>
    <nav class="step-navigation"><a role="button" class="pagination-previous">BACK</a> <a role="button" class="pagination-next disabled">NEXT STEP</a></nav>
  </div>
  <div class="step-container AD inactive enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Attribution Details
      </h5>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`Stepper.vue renders correctly has expected UI on ND step 1`] = `
<div class="stepper-container column">
  <div class="step-container FS previous enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        License Expertise
      </h5>
    </div>
    <firststep-stub stepid="0" status="previous"></firststep-stub>
    <!---->
  </div>
  <div class="step-container BY previous enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Attribution
      </h5>
    </div>
    <step-stub stepname="BY" stepid="1" status="previous" enabled="true" selected="true"></step-stub>
    <!---->
  </div>
  <div class="step-container NC previous enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Commercial Use
      </h5>
    </div>
    <step-stub stepname="NC" stepid="2" status="previous" reversed="true" enabled="true"></step-stub>
    <!---->
  </div>
  <div class="step-container ND current enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Do you want to allow others to remix, adapt, or build upon your work?
      </h5>
    </div>
    <step-stub stepname="ND" stepid="3" status="current" reversed="true" enabled="true"></step-stub>
    <nav class="step-navigation"><a role="button" class="pagination-previous">BACK</a> <a role="button" class="pagination-next disabled">NEXT STEP</a></nav>
  </div>
  <div class="step-container SA inactive enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Sharing Requirements
      </h5>
    </div>
    <!---->
    <!---->
  </div>
  <div class="step-container AD inactive enabled">
    <div class="step-header">
      <h5 class="step-title vocab hb h5b">
        Attribution Details
      </h5>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;
